File: DFS/트리의_부모_찾기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
    static List<List<Integer>> graph;
    static int[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        
        graph = new ArrayList<>();
        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int i = 0; i < N - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        visited = new int[N + 1];
        dfs(1);
        
        StringBuilder sb = new StringBuilder();
        for (int i = 2; i <= N; i++) {
            sb.append(visited[i]).append("\n");
        }
        System.out.print(sb.toString());
    }
    
    static void dfs(int parent) {
        for (int child : graph.get(parent)) {
            if (visited[child] != 0) { // If already visited, skip
                continue;
            }
            visited[child] = parent;
            dfs(child);
        }
    }
}

File: DFS/트리의_부모_찾기.py
import sys
sys.setrecursionlimit(10**6)
N = int(input())
 
graph = [[] for _ in range(N+1)]
for i in range(N-1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
 
visited = [0]*(N+1) 
 
def dfs(parent):
    for child in graph[parent]: 
        if visited[child] != 0: # 방문한 적이 있다면
            continue
        visited[child] = parent
        dfs(child)
 
dfs(1)
 
for i in range(2, N+1):
    print(visited[i])
